### PR TITLE
feat(zkevm): add test for execution witness bytecodes for contract creation, CALL-variants and empty block

### PR DESCRIPTION
**Pull request by jsign**
## 🗒️ Description
This PR adds more test coverage for `ExecutionWitness.codes` cases, in three main groups:
- Empty block: in this case we do a strict check that only system contracts bytecodes are present (and nothing else).
- CALL-like opcodes: for multiple CALL-variants check that target bytecodes work as expected.
- Contract creation: test contract creation via tx creation, CREATE and CREATE2.

This PR isn't targeting full coverage of codes in witness, more PRs will come soon. Just opening PRs in parts to avoid them being too big and be easy to review.

```
> uv run fill --clean -m "blockchain_test" --fork Amsterdam ./tests/amsterdam/eip8025_optional_proofs/
========================================================== test session starts ==========================================================
platform linux -- Python 3.12.3, pytest-8.4.2, pluggy-1.6.0
Generating fixtures for: Amsterdam, BPO2ToAmsterdamAtTime15k
2.19.0
Log file: logs/fill-20260310-222215-main.log
rootdir: /home/ignacio/code/execution-specs
configfile: packages/testing/src/execution_testing/cli/pytest_commands/pytest_ini_files/pytest-fill.ini
plugins: cov-4.1.0, metadata-3.1.1, json-report-1.5.0, xdist-3.8.0, regex-0.2.0, custom-report-1.0.1, html-4.1.1
collecting ...  pytest-regex selected 50 tests to run for regex: .*
collected 75 items / 25 deselected / 50 selected                                                                                        

tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_call_variants.py .............                                    [13/25]
tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py ...........                                  [24/25]
tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py .                                             [25/25]

=========================================================== warnings summary ============================================================
.venv/lib/python3.12/site-packages/_pytest/config/__init__.py:812
  /home/ignacio/code/execution-specs/.venv/lib/python3.12/site-packages/_pytest/config/__init__.py:812: PytestAssertRewriteWarning: Module already imported so cannot be rewritten; execution_testing.cli.pytest_commands.plugins.filler.pre_alloc
    self.import_plugin(arg, consider_entry_points=True)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---------------------------- Log file: /home/ignacio/code/execution-specs/logs/fill-20260310-222215-main.log ----------------------------
------------------- Generated html report: file:///home/ignacio/code/execution-specs/fixtures/.meta/report_fill.html --------------------
=====================================  T8n cache: 25 unique test groups, no cache sharing possible ======================================
======================  No tests executed - the test fixtures in "fixtures/" may now be executed against a client  ======================
============================================= 25 passed, 25 deselected, 1 warning in 1.45s ==============================================
```

## 🔗 Related Issues or PRs
https://github.com/ethereum/execution-specs/issues/2183

## ✅ Checklist

- [ ] All: Ran fast `tox` checks to avoid unnecessary CI fails, see also [Code Standards](https://eest.ethereum.org/main/getting_started/code_standards/) and [Enabling Pre-commit Checks](https://eest.ethereum.org/main/dev/precommit/):
    ```console
    uvx tox -e static
    ```
- [ ] All: PR title adheres to the [repo standard](https://eest.ethereum.org/main/getting_started/contributing/?h=contri#commit-messages-issue-and-pr-titles) - it will be used as the squash commit message and should start `type(scope):`.
- [ ] All: Considered updating the online docs in the [./docs/](/ethereum/execution-specs/blob/HEAD/docs/) directory.
- [ ] All: Set appropriate labels for the changes (only maintainers can apply labels).
- [ ] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://eest.ethereum.org/main/tests/) are correctly formatted.
- [ ] Tests: For PRs implementing a missed test case, update the [post-mortem document](/ethereum/execution-specs/blob/HEAD/docs/writing_tests/post_mortems.md) to add an entry the list.
- [ ] Ported Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) or [tests/static](/ethereum/execution-specs/blob/HEAD/tests/static) have been assigned `@ported_from` marker.

#### Cute Animal Picture

Put a link to a cute animal picture inside the parenthesis-->

### CODE REVIEW

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes.py
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes.py:1]: The two existing tests were moved to tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py.py to have better grouping of tests sub-domains.

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_call_variants.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        pytest.param(Op.CALLCODE, id="callcode"),
+        pytest.param(Op.STATICCALL, id="staticcall"),
+    ],
+)
+def test_witness_codes_call_existing_contract(
+    pre: Alloc,
+    blockchain_test: BlockchainTestFiller,
+    call_opcode: Op,
+) -> None:
+    """
```
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_call_variants.py:35]: Hopefully each test description is clear enough to explain what is going on, so I'll avoid re-explaining in comments. I might still add some extra comments if worth it.

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_call_variants.py [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+                            Bytes(bytes(code_c)),
+                        ],
+                    )
+                ),
+            )
+        ],
+        post={
+            sender: Account(nonce=1),
+        },
+    )
```
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_call_variants.py:113]: I wanted to add the OOG cases I raised in Discord -- but let's wait until resolving how code is pulled.

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    The pre-state contract's bytecode should appear in
+    executionWitness.codes because get_witness_codes checks per-address
+    pre-state match. The pre-state contract qualifies regardless of the
+    earlier CREATE writing the same hash.
+
+    TODO(zkevm): it might be worth reconsidering how the execution witness
+    generation handles this case, since technically this same bytecode was
+    observed in tx1 thus can in theory be exploited to avoid including it
+    for tx2. Feels inconsistent with the other test
+    test_witness_codes_create_then_call_same_block(...).
```
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py:305]: Relevant comment: let me repeat a bit what is going on here to be clear:
- The test above `test_witness_codes_create_then_call_same_block(...)` says that if in a tx1 I create contract with addr A, and in tx2 I call addr A, then the bytecode won't appear since in both txs we're targeting the same addr. The way the tracker works tracks bytecode per addr.
- In this case, tx1 creates a contract with addr A and codehash codeHash1. In tx2 we call another addr B with the same codeHash. In this case the bytecode appears.

This feels inconsistent since the situation is the same (i.e. the CALL calls some bytecode created in a prev tx) but the only reason why in one case appears or doesn't appear in the witness is that the target addr is the same.

For now keeping the test to be consistent with how the spec works -- but feels to me we might want to reconsider this. 

Open to feedback

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    block (no user transactions) should have exactly the four system
+    contract bytecodes in executionWitness.codes and nothing else.
+    """
+    alloc = fork.pre_allocation_blockchain()
+
+    system_codes = []
+    for addr in SYSTEM_CONTRACT_ADDRS:
+        code = alloc[addr]["code"]
+        if isinstance(code, str):
+            code = bytes.fromhex(code.removeprefix("0x"))
```
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py:46]: I was a bit surprised to find that system contracts bytecode returned by `pre_allocation_blockchain` can have different types. Apparently, for the Cancun system contracts, it returns a `str` and for Osaka they are read from a .bin file, thus have other type.

NBD, but maybe there is some homogenization opportunity that can be done.

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+    blockchain_test(
+        pre=pre,
+        blocks=[
+            Block(
+                txs=[],
+                expected_execution_witness_codes=(
+                    ExecutionWitnessCodesExpectation(
+                        codes_present=system_codes,
+                        allow_unexpected=False,
```
Comment (jsign) [inline tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py:57]: Note the strict check here.

**Comment by jsign** [comment]
cc @kevaundray ready for some eyes without rush.

**Review by kevaundray — APPROVED** [review]

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py (hunk 3 above)
Comment (kevaundray) [reply tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py:305]: unless I misunderstood, it seems that the bytecode won't be needed in the execution witness since its in tx1?

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py (hunk 4 above)
Comment (kevaundray) [reply tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_system_contracts.py:46]: cc maybe @gurukamath for visibility -- we could create an issue since its orthogonal to this PR

File: tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py (hunk 3 above)
Comment (jsign) [reply tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py:305]: Yep -- the logic is that the execution of a previous transaction generated some bytecode that will match the code_hash of a future tx. Depending if both txs target the same or different accounts, we have different behaviours today -- which I think is the thing we can polish better in the spec to behave the same (i.e. do not include in the witness). I'll look into this soon.

Comment (jsign) [reply tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py:305]: Uhm, I think this case might be more hairy than we think in the sense that it might not be able to resolve everything so elegantly.

I'll look further and explain better in a PR that attempts a fix and we can see the implications in tests. Will be easier to see by looking at code I think.

Comment (jsign) [reply tests/amsterdam/eip8025_optional_proofs/test_witness_bytecodes_contract_creation.py:305]: https://github.com/ethereum/execution-specs/pull/2479
